File: GravitySort.py
import time
#Loon suvalise listi, n on listi pikkus, k on suvaliste arvude vahemik
import random
randomList = []
n = 10000
k = 10
for i in range(n):
    randomList.append(random.randint(0,k))

def gravitatsiooniSortimine(arr):
    pikkus = len(arr)
    suurim = max(arr)
    
    nupud = [[0]*suurim for j in range(pikkus)]
    
    for i in range(pikkus):
        for j in range(arr[i]):
            nupud[i][j] = 1
    
    for n in range(suurim):
        kogus = 0
        for i in range(pikkus):
            kogus += nupud[i][n]
            nupud[i][n] = 0
        for i in range(pikkus-1, pikkus-kogus-1, -1):
            nupud[i][n] = 1


    output = [0]*pikkus
    for i in range(pikkus):
        output[i] = nupud[i].count(1)
    return(output)
        
    

start = time.time()
gravitatsiooniSortimine(randomList)
print(time.time()-start)

File: gravitySort.cpp
#include <iostream>
#include <cstdlib>
#include <ctime>
#include <array>
#include <chrono>
#include <vector>


using namespace std;
using namespace std::chrono;

void printArray(vector<int> arr, int n)
{
    int i;
    for (i = 0; i < n; i++)
        cout << arr[i] << " ";
    cout << endl;
}

void gravitatsiooniSortimine(vector<int> &arr) {
    int len = arr.size();
    int suurim = *max_element(arr.begin(), arr.end());
    
    vector<vector<int>> nupud(len, vector<int>(suurim, 0));
    for (int i = 0; i < len; ++i){
        for (int j = 0; j < arr[i] ; j++) {
            nupud[i][j] = 1;
        }
    }

    for (int i = 0; i <suurim ; i++) {
        int kogus =0;
        for (int j = 0; j < len; j++) {
            kogus += nupud[j][i];
            nupud[j][i] = 0;
        }
        for (int j = len-kogus; j < len; j++){
            nupud[j][i] = 1;
        }
    }
    vector<int> output(len, 0);
    for (int i = 0; i<len; i++){
        
        output[i] = count(nupud[i].begin(), nupud[i].end(), 1);
    }
    arr = output;
}

int main() {
    
    const int n = 10000;
    const int k = 10;
    vector<int> vect;

    srand(time(NULL));
    for (int i = 0; i < n; ++i){
        
        vect.push_back(rand() %k) ;
    }
    //printArray(vect, n);

    auto start = high_resolution_clock::now();
    gravitatsiooniSortimine(vect);
    auto stop = high_resolution_clock::now();
    auto duration = duration_cast<microseconds>(stop - start);
    cout << endl;
    cout << duration.count() << " microseconds" << endl;
    //printArray(vect, n);
    return 0;
}